libraries/standard/mqtt/src/iot_mqtt_operation.c

```diff
@@ -436,6 +463,107 @@ static bool _completePendingSend( _mqttOperation_t * pOperation,
+     * wait on. */
+    if( waitable == true )
+    {
+        /* Create a semaphore to wait on for a waitable operation. */
+        if( IotSemaphore_Create( &( pOperation->u.operation.notify.waitSemaphore ), 0, 1 ) == false )
+        {
+            IotLogError( "(MQTT connection %p) Failed to create semaphore for "
+                         "waitable operation.",
+                         pMqttConnection );
+
+            status = IOT_MQTT_NO_MEMORY;
+        }
+        else
+        {
+            /* A waitable operation is created with an additional reference for the
+             * Wait function. */
+            ( pOperation->u.operation.jobReference )++;
+        }
+    }
+    else
+    {
+        /* If the waitable flag isn't set but a callback is, copy the callback
+         * information. */
+        if( pCallbackInfo != NULL )
+        {
+            pOperation->u.operation.notify.callback = *pCallbackInfo;
+        }
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static bool _sendPingRequest( _mqttConnection_t * pMqttConnection,
+                              _mqttOperation_t * pPingreqOperation )
+{
+    size_t bytesSent = 0;
+    bool status = true;
+    uint32_t swapStatus = 0;
+
+    IotMqtt_Assert( pMqttConnection != NULL );
+    IotMqtt_Assert( pPingreqOperation != NULL );
+
+    IotLogDebug( "(MQTT connection %p) Sending PINGREQ.", pMqttConnection );
+
+    /* Because PINGREQ may be used to keep the MQTT connection alive, it is
+     * more important than other operations. Bypass the queue of jobs for
+     * operations by directly sending the PINGREQ in this job. */
+    bytesSent = pMqttConnection->pNetworkInterface->send( pMqttConnection->pNetworkConnection,
+                                                          pPingreqOperation->u.operation.pMqttPacket,
+                                                          pPingreqOperation->u.operation.packetSize );
+
+    if( bytesSent != pPingreqOperation->u.operation.packetSize )
+    {
+        IotLogError( "(MQTT connection %p) Failed to send PINGREQ.", pMqttConnection );
+        status = false;
+    }
+    else
+    {
+        /* Assume the keep-alive will fail. The network receive callback will
+         * clear the failure flag upon receiving a PINGRESP. */
+        swapStatus = Atomic_CompareAndSwap_u32( &( pPingreqOperation->u.operation.periodic.ping.failure ),
+                                                1,
+                                                0 );
+        IotMqtt_Assert( swapStatus == 1U );
+
+        /* Set the period for scheduling a PINGRESP check. */
+        pPingreqOperation->u.operation.periodic.ping.nextPeriodMs = IOT_MQTT_RESPONSE_WAIT_MS;
+
+        IotLogDebug( "(MQTT connection %p) PINGREQ sent. Scheduling check for PINGRESP in %d ms.",
+                     pMqttConnection,
+                     IOT_MQTT_RESPONSE_WAIT_MS );
```

Comment:
A comment was made in #700 that this log message is misplaced since the scheduling does not occur in this function. Since line 887 also logs the next scheduled keep-alive job, the reference to scheduling here can just be omitted.